mesos-client/src/test/scala/com/mesosphere/mesos/client/SessionActorTest.scala

```diff
@@ -0,0 +1,124 @@
+package com.mesosphere.mesos.client
+
+import akka.pattern.ask
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.HttpCredentials
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import akka.testkit.TestProbe
+import akka.util.Timeout
+import com.mesosphere.utils.AkkaUnitTest
+
+import scala.concurrent.duration._
+
+class SessionActorTest extends AkkaUnitTest {
+
+  "The Session Actor" should {
+    "replied with a response to the original sender" in {
+      val f = new Fixture
+
+      Given("A SessionActor and an HTTP response")
+      val sessionActor =
+        system.actorOf(SessionActor.props(BasicAuthenticationProvider("user", "password"), f.requestFactory))
+      val originalSender = TestProbe()
+      val httpResponse = HttpResponse(entity = HttpEntity("hello"))
+
+      When("the actor receives a response for the request")
+      sessionActor ! SessionActor.Response(Array.emptyByteArray, originalSender.ref, httpResponse)
+
+      Then("it sends the response to the original sender")
+      originalSender.expectMsg(httpResponse)
+    }
+
+    "make a simple call to Mesos" in withMesosStub(StatusCodes.OK) { mesos =>
+      implicit val timeout = Timeout(2.seconds)
+
+      Given("A simple Mesos API stub and a SessionActor instance")
+      val sessionActor =
+        system.actorOf(SessionActor.props(BasicAuthenticationProvider("user", "password"), mesos.createRequest))
+
+      When("we make a call through the session actor")
+      val response = (sessionActor ? Array.empty[Byte]).futureValue.asInstanceOf[HttpResponse]
+
+      Then("we receive an HTTP 200")
+      response.status should be(StatusCodes.OK)
+
+      And("Mesos was called only once")
+      Unmarshal(response.entity).to[String].futureValue.toInt should be(1)
+    }
+
+    "refresh the session token when the first call response is an HTTP 401" in withMesosStub(
+      StatusCodes.Unauthorized,
+      StatusCodes.OK) { mesos =>
+      implicit val timeout = Timeout(2.seconds)
+
+      Given("A SessionActor instance")
+      val sessionActor =
+        system.actorOf(SessionActor.props(BasicAuthenticationProvider("user", "password"), mesos.createRequest))
+
+      When("we make a call through the session actor")
+      val response = (sessionActor ? Array.empty[Byte]).futureValue.asInstanceOf[HttpResponse]
+
+      Then("we receive an HTTP 200")
+      response.status should be(StatusCodes.OK)
+
+      And("Mesos was called only once")
+      Unmarshal(response.entity).to[String].futureValue.toInt should be(2)
```

Comment:
It's somewhat paranoid but can we make sure that the credentials provider has indeed issued a second token?